eZ/Publish/Core/Persistence/Legacy/Content/Type/Gateway/DoctrineDatabase.php

```diff
@@ -88,275 +91,282 @@ class DoctrineDatabase extends Gateway
                 )
-                ->set('language_locale', $query->bindValue($language))
-                ->set('name', $query->bindValue($name));
-            $query->prepare()->execute();
+            );
         }
     }
 
-    /**
-     * Inserts a new content type.
-     *
-     * @param \eZ\Publish\SPI\Persistence\Content\Type $type
-     * @param mixed|null $typeId
-     *
-     * @return mixed Type ID
-     */
-    public function insertType(Type $type, $typeId = null)
+    public function insertType(Type $type, ?int $typeId = null): int
     {
-        $q = $this->dbHandler->createInsertQuery();
-        $q->insertInto($this->dbHandler->quoteTable('ezcontentclass'));
-        $q->set(
-            $this->dbHandler->quoteColumn('id'),
-            isset($typeId) ? $q->bindValue($typeId, null, \PDO::PARAM_INT) : $this->dbHandler->getAutoIncrementValue('ezcontentclass', 'id')
-        )->set(
-            $this->dbHandler->quoteColumn('version'),
-            $q->bindValue($type->status, null, \PDO::PARAM_INT)
-        )->set(
-            $this->dbHandler->quoteColumn('created'),
-            $q->bindValue($type->created, null, \PDO::PARAM_INT)
-        )->set(
-            $this->dbHandler->quoteColumn('creator_id'),
-            $q->bindValue($type->creatorId, null, \PDO::PARAM_INT)
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->insert(self::CONTENT_TYPE_TABLE)
+            ->values(
+                [
+                    // Legacy Storage: "version" stores CT status (draft, published)
+                    'version' => $query->createNamedParameter(
+                        $type->status,
+                        ParameterType::INTEGER,
+                        ':status'
+                    ),
+                    'created' => $query->createNamedParameter(
+                        $type->created,
+                        ParameterType::INTEGER,
+                        ':created'
+                    ),
+                    'creator_id' => $query->createNamedParameter(
+                        $type->creatorId,
+                        ParameterType::INTEGER,
+                        ':creator_id'
+                    ),
+                ]
+            );
+        $this->setNextAutoIncrementedValueIfAvailable(
+            $query,
+            self::CONTENT_TYPE_TABLE,
+            'id',
+            self::CONTENT_TYPE_SEQ,
+            $typeId
         );
-        $this->setCommonTypeColumns($q, $type);
 
-        $q->prepare()->execute();
+        $columnQueryValueAndTypeMap = $this->mapCommonContentTypeColumnsToQueryValuesAndTypes(
+            $type
+        );
+        foreach ($columnQueryValueAndTypeMap as $columnName => $data) {
+            [$value, $parameterType] = $data;
+            $query
+                ->setValue(
+                    $columnName,
+                    $query->createNamedParameter($value, $parameterType, ":{$columnName}")
+                );
+        }
```

Comment:
This logic is duplicated in 3 other places.

Reply:
> This logic is duplicated in 3 other places.

Actually it is not. `insert` uses either `values` or `setValue` methods, while `update` uses `set` method. I'm open to suggestions as to refactoring of this, so it's clear this code is unique. Anything except except using `$methodName = 'setValue'`, because it would be a maintenance /refactoring / usage search nightmare :stuck_out_tongue_winking_eye:

Reply:
> Actually it is not. insert uses either values or setValue methods, while update uses set method

You are right, I overlooked this.